dff/core/pipeline/messenger_interface.py

```diff
@@ -0,0 +1,148 @@
+        By default, it logs all exit exceptions to `info` log and all non-exit exceptions to `error`.
+        :e: - the exception.
+        """
+        if isinstance(e, Exception):
+            logger.error(f"Exception in {type(self).__name__} loop!\n{str(e)}")
+        else:
+            logger.info(f"{type(self).__name__} has stopped polling.")
+
+    async def connect(
+        self,
+        pipeline_runner: PipelineRunnerFunction,
+        loop: PollingProviderLoopFunction = lambda: True,
+        timeout: int = 0,
+    ):
+        """
+        Method, running a request - response cycle in a loop.
+        The looping behaviour is determined by :loop: and :timeout:,
+            for most cases the loop itself shouldn't be overridden.
+        :loop: - a function that determines whether polling should be continued;
+            called in each cycle, should return True to continue polling or False to stop.
+        :timeout: - a time interval between polls (in seconds).
+        """
+        while loop():
+            try:
+                user_updates = self._request()
+                responses = [await pipeline_runner(request, ctx_id) for request, ctx_id in user_updates]
+                self._respond(responses)
+                await asyncio.sleep(timeout)
+
+            except BaseException as e:
+                self._on_exception(e)
+                break
+
+
+class CallbackMessengerInterface(MessengerInterface):
+    """
+    Callback message interface is waiting for user input and answers once it gets one.
+    """
+
+    def __init__(self):
+        self._pipeline_runner: Optional[PipelineRunnerFunction] = None
+
+    async def connect(self, pipeline_runner: PipelineRunnerFunction):
+        self._pipeline_runner = pipeline_runner
+
+    def on_request(self, request: Any, ctx_id: Hashable) -> Context:
+        """
+        Method invoked on user input.
+        This method works just like `Pipeline.__call__(request, ctx_id)`,
+            however callback message interface may contain additional functionality (e.g. for external API accessing).
+        :request: - user input.
+        :ctx_id: - any unique id that will be associated with dialog between this user and pipeline.
+        Returns context that represents dialog with the user;
+            `last_response`, `id` and some dialog info can be extracted from there.
+        """
+        return asyncio.run(self._pipeline_runner(request, ctx_id))
+
+
+class CLIMessengerInterface(PollingMessengerInterface):
+    """
+    Command line message interface - the default message interface, communicating with user via STDIN/STDOUT.
+    This message interface can maintain dialog with one user at a time only.
+    """
+
+    def __init__(
+        self,
+        intro: Optional[str] = None,
+        prompt_request: str = "request: ",
+        prompt_response: str = "response: ",
+        out_descriptor: TextIO = None,
```

Comment:
I would love actually if any static Python analyzer complained in case of assigning non-`Optional`-type-hinted variable to None. I mean why
1. `var: str = None` is OK
2. `var: str = 7` is not OK
They're both wrong!